client/castApp/src/castinit.c

```diff
@@ -65,13 +67,133 @@ static void casthook(initHookState state)
+        /* Attempt to add after iocInit(), when we may be connected.
+           To fully support, would need to force reconnect or resend w/ updated envs list. */
+        errlogSevPrintf(errlogMinor, "addReccasterEnvVars called after iocInit() when reccaster might already be connected. Not supported\n");
+        epicsMutexUnlock(self->lock);
+        return;
+    }
+    const int num_new_extra_envs = self->num_extra_envs + argc;
+    char **new_extra_envs = calloc(num_new_extra_envs, sizeof(*new_extra_envs));
+    if(new_extra_envs == NULL) {
+        errlogSevPrintf(errlogMajor, "Error in memory allocation of new_extra_envs from addReccasterEnvVars\n");
+        epicsMutexUnlock(self->lock);
+        return;
+    }
+    /* copy self->extra_envs into new_extra_envs with room for new envs */
+    for(i=0; i < self->num_extra_envs; i++) {
+        if((new_extra_envs[i] = strdup(self->extra_envs[i])) == NULL) {
+            errlogSevPrintf(errlogMinor, "strdup error from addReccasterEnvVars\n", new_extra_envs[j]);
+            ret = 1;
+            break;
+        }
+    }
+    int found_dup;
+    int bad_arg_count = 0;
+    /* sanitize input - check for dups and empty args */
+    if(!ret) {
+        for(i=0, j=self->num_extra_envs; i < argc; i++) {
+            if(argv[i] == NULL) {
+                bad_arg_count++;
+                errlogSevPrintf(errlogMinor, "Arg is NULL for addReccasterEnvVars\n");
+                continue;
+            }
+            else if(argv[i][0] == '\0') {
+                bad_arg_count++;
+                errlogSevPrintf(errlogMinor, "Arg is empty for addReccasterEnvVars\n");
+                continue;
+            }
+            found_dup = 0;
+            /* check if dup in self->extra_envs. doesn't check if arg is in default_envs right now */
+            for(k = 0; k < num_new_extra_envs && new_extra_envs[k]; k++) {
+                if(strcmp(argv[i], new_extra_envs[k]) == 0) {
+                    found_dup = 1;
+                    errlogSevPrintf(errlogMinor, "Env var %s is already in extra_envs list\n", argv[i]);
+                    break;
+                }
+            }
+            if(found_dup) {
+                bad_arg_count++;
+                continue;
+            }
+            if((new_extra_envs[j] = strdup(argv[i])) == NULL) {
+                errlogSevPrintf(errlogMinor, "strdup error for copying %s to new_extra_envs[%d] from addReccasterEnvVars\n", new_extra_envs[j]);
+                ret = 1;
+                break;
+            }
+            /* this is a valid arg and we have added the new env var to our array, increment new_extra_envs index */
+            j++;
+        }
+    }
+    /* if we have no allocation issues and have at least one new env var that is valid, add to self->extra_envs */
+    if(!ret && bad_arg_count != argc) {
+        char ** new_envs;
+        new_envs = realloc(self->extra_envs, sizeof(* new_envs) * (self->num_extra_envs + (argc-bad_arg_count)));
```

Comment:
Why not replace `self->extra_envs` with `new_extra_envs` here instead of another alloc+copy?  In the common case (no duplicates) they will be identical, and in the uncommon case only a few bytes are wasted.